File: EASYTelegramBot/DomandaDAO.py
import sqlite3

from Domanda import Domanda


class DomandaDAO:
    def __init__(self, db_manager):
        self.db_manager = db_manager

    async def do_retrieve_all(self):
        try:
            async with self.db_manager as conn:
                cursor = await conn.execute("SELECT * FROM Domande")
                rows = await cursor.fetchall()
                domande = []
                for row in rows:
                    domande.append(
                        Domanda(row[0], row[1], row[2], row[3], row[4], row[5], row[6], row[7], row[8], row[9],
                                row[10], row[11]))
                return domande
        except sqlite3.Error as e:
            print(f"Errore durante la ricerca di tutte le domande: {e}")

    async def do_retrieve_by_id(self, numeroDomanda):
        try:
            async with self.db_manager as conn:
                cursor = await conn.execute("SELECT * FROM Domande WHERE numeroDomanda = ?", (numeroDomanda,))
                row = await cursor.fetchone()
                if row:
                    return Domanda(row[0], row[1], row[2], row[3], row[4], row[5], row[6], row[7], row[8], row[9],
                                   row[10], row[11])
                return None
        except sqlite3.Error as e:
            print(f"Errore durante la ricerca della domanda per Numero: {e}")

    async def do_retrieve_by_argomento(self, argomento):
        try:
            async with self.db_manager as conn:
                cursor = await conn.execute("SELECT * FROM Domande WHERE argomento = ?", (argomento,))
                rows = await cursor.fetchall()
                domande = []
                for row in rows:
                    domande.append(
                        Domanda(row[0], row[1], row[2], row[3], row[4], row[5], row[6], row[7], row[8], row[9],
                                row[10], row[11]))
                return domande
        except sqlite3.Error as e:
            print(f"Errore durante la ricerca delle domande per Argomento: {e}")

    async def do_retrieve_by_argomento_random_limit(self, argomento, limite):
        try:
            async with self.db_manager as conn:
                cursor = await conn.execute("SELECT * FROM Domande WHERE argomento = ? ORDER BY RANDOM() LIMIT ?", (argomento, limite))
                rows = await cursor.fetchall()
                domande = []
                for row in rows:
                    domande.append(
                        Domanda(row[0], row[1], row[2], row[3], row[4], row[5], row[6], row[7], row[8], row[9],
                                row[10], row[11]))
                return domande
        except sqlite3.Error as e:
            print(f"Errore durante la ricerca delle domande per Argomento: {e}")

    async def do_save(self, domanda):
        try:
            async with self.db_manager as conn:
                await conn.execute(
                    "INSERT INTO Domande (testo, argomento, rispostaA, rispostaB, rispostaC, rispostaD, rispostaCorretta, difficolta, tempoRisposta, meme, fonte) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
                    (domanda.testo, domanda.argomento, domanda.rispostaA, domanda.rispostaB, domanda.rispostaC,
                     domanda.rispostaD, domanda.rispostaCorretta, domanda.difficolta, domanda.tempoRisposta,
                     domanda.meme, domanda.fonte))
                await conn.commit()
        except sqlite3.Error as e:
            print(f"Errore durante l'inserimento della domanda: {e}")

    async def do_update(self, domanda):
        try:
            async with self.db_manager as conn:
                await conn.execute(
                    "UPDATE Domande SET testo = ?, argomento = ?, rispostaA = ?, rispostaB = ?, rispostaC = ?, rispostaD = ?, rispostaCorretta = ?, difficolta = ?, tempoRisposta = ?, meme = ?, fonte = ? WHERE numeroDomanda = ?",
                    (domanda.testo, domanda.argomento, domanda.rispostaA, domanda.rispostaB, domanda.rispostaC,
                     domanda.rispostaD, domanda.rispostaCorretta, domanda.difficolta, domanda.tempoRisposta,
                     domanda.meme, domanda.fonte, domanda.numeroDomanda))
                await conn.commit()
        except sqlite3.Error as e:
            print(f"Errore durante l'aggiornamento della domanda: {e}")

    async def do_save_or_update(self, domanda):
        if await self.do_retrieve_by_id(domanda.numeroDomanda):
            await self.do_update(domanda)
        else:
            await self.do_save(domanda)

    async def do_delete(self, numeroDomanda):
        try:
            async with self.db_manager as conn:
                await conn.execute("DELETE FROM Domande WHERE numeroDomanda = ?", (numeroDomanda,))
                await conn.commit()
        except sqlite3.Error as e:
            print(f"Errore durante l'eliminazione della domanda: {e}")

File: ImportExportDomandeAutomatico/ImportExportDomandeAutomatico/Domande/DomandaDAO.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Domande;
using Microsoft.Data.Sqlite;

namespace Domande
{
    public class DomandaDAO : Storage.SQLDAO
    {
        public DomandaDAO(SqliteConnection connection) : base(connection)
        {
        }

        public List<Domanda> DoRetrieveAll()
        {
            Debug.WriteLine("DoRetrieveAll");
            List<Domanda> domande = new List<Domanda>();
            SqliteCommand command = new SqliteCommand("SELECT * FROM Domande", connection);
            SqliteDataReader reader = command.ExecuteReader();
            while (reader.Read())
            {
                domande.Add(new Domanda(reader.GetInt32(0), reader.GetString(1), reader.GetString(2), reader.GetString(3), reader.GetString(4), reader.GetString(5), reader.GetString(6), reader.GetInt32(7), reader.GetInt32(8), reader.GetInt32(9), reader.IsDBNull(10) ? null : reader.GetString(10), reader.IsDBNull(11) ? null : reader.GetString(11)));
            }
            return domande;
        }

        public Domanda DoRetrieveById(int id)
        {
            Debug.WriteLine("DoRetrieveById");
            SqliteCommand command = new SqliteCommand("SELECT * FROM Domande WHERE NumeroDomanda = @id", connection);
            command.Parameters.AddWithValue("@id", id);
            SqliteDataReader reader = command.ExecuteReader();
            if (reader.Read())
            {
                return new Domanda(reader.GetInt32(0), reader.GetString(1), reader.GetString(2), reader.GetString(3), reader.GetString(4), reader.GetString(5), reader.GetString(6), reader.GetInt32(7), reader.GetInt32(8), reader.GetInt32(9), reader.IsDBNull(10) ? null : reader.GetString(10), reader.IsDBNull(11) ? null : reader.GetString(11));
            }
            return null;
        }

        public List<Domanda> DoRetrieveByArgomento(string argomento)
        {
            Debug.WriteLine("DoRetrieveByArgomento");
            List<Domanda> domande = new List<Domanda>();
            SqliteCommand command = new SqliteCommand("SELECT * FROM Domande WHERE Argomento = @argomento", connection);
            command.Parameters.AddWithValue("@argomento", argomento);
            SqliteDataReader reader = command.ExecuteReader();
            while (reader.Read())
            {
                domande.Add(new Domanda(reader.GetInt32(0), reader.GetString(1), reader.GetString(2), reader.GetString(3), reader.GetString(4), reader.GetString(5), reader.GetString(6), reader.GetInt32(7), reader.GetInt32(8), reader.GetInt32(9), reader.IsDBNull(10) ? null : reader.GetString(10), reader.IsDBNull(11) ? null : reader.GetString(11)));
            }
            return domande;
        }

        public List<Domanda> DoRetrieveByTesto(string testo)
        {
            Debug.WriteLine("DoRetrieveByTesto");
            List<Domanda> domande = new List<Domanda>();
            SqliteCommand command = new SqliteCommand("SELECT * FROM Domande WHERE Testo LIKE @testo", connection);
            command.Parameters.AddWithValue("@testo", "%" + testo + "%"); // Aggiunge % al testo per cercare tutte le sottostringhe
            SqliteDataReader reader = command.ExecuteReader();
            while (reader.Read())
            {
                domande.Add(new Domanda(reader.GetInt32(0), reader.GetString(1), reader.GetString(2), reader.GetString(3), reader.GetString(4), reader.GetString(5), reader.GetString(6), reader.GetInt32(7), reader.GetInt32(8), reader.GetInt32(9), reader.IsDBNull(10) ? null : reader.GetString(10), reader.IsDBNull(11) ? null : reader.GetString(11)));
            }
            // Restituisci l'elenco delle domande trovate
            return domande;
        }



        public List<Domanda> DoRetrieveByDifficolta(int difficolta)
        {
            Debug.WriteLine("DoRetrieveByDifficolta");
            List<Domanda> domande = new List<Domanda>();
            SqliteCommand command = new SqliteCommand("SELECT * FROM Domande WHERE Difficolta = @difficolta", connection);
            command.Parameters.AddWithValue("@difficolta", difficolta);
            SqliteDataReader reader = command.ExecuteReader();
            while (reader.Read())
            {
                domande.Add(new Domanda(reader.GetInt32(0), reader.GetString(1), reader.GetString(2), reader.GetString(3), reader.GetString(4), reader.GetString(5), reader.GetString(6), reader.GetInt32(7), reader.GetInt32(8), reader.GetInt32(9), reader.IsDBNull(10) ? null : reader.GetString(10), reader.IsDBNull(11) ? null : reader.GetString(11)));
            }
            return domande;
        }

        public bool DoSave(Domanda domanda)
        {
            Debug.WriteLine("DoSave");
            SqliteCommand command = new SqliteCommand(
                @"INSERT INTO Domande (testo, argomento, RispostaA, RispostaB, RispostaC, RispostaD, 
                RispostaCorretta, difficolta, tempoRisposta, meme, fonte) 
                VALUES (@testo, @argomento, @RispostaA, @RispostaB, @RispostaC, @RispostaD, 
                @RispostaCorretta, @difficolta, @tempoRisposta, @meme, @fonte)", connection);

            command.Parameters.AddWithValue("@testo", domanda.Testo);
            command.Parameters.AddWithValue("@argomento", domanda.Argomento);
            command.Parameters.AddWithValue("@RispostaA", domanda.RispostaA);
            command.Parameters.AddWithValue("@RispostaB", domanda.RispostaB);
            command.Parameters.AddWithValue("@RispostaC", domanda.RispostaC);
            command.Parameters.AddWithValue("@RispostaD", domanda.RispostaD);
            command.Parameters.AddWithValue("@RispostaCorretta", domanda.RispostaCorretta);
            command.Parameters.AddWithValue("@difficolta", domanda.Difficolta);
            command.Parameters.AddWithValue("@tempoRisposta", domanda.TempoRisposta);
            command.Parameters.AddWithValue("@meme", domanda.Meme is null ? DBNull.Value : (object)domanda.Meme);
            command.Parameters.AddWithValue("@fonte", domanda.Fonte is null ? DBNull.Value : (object)domanda.Fonte);

            return command.ExecuteNonQuery() == 1;
        }

        public bool DoUpdate(Domanda domanda)
        {
            Debug.WriteLine("DoUpdate");
            SqliteCommand command = new SqliteCommand(
                @"UPDATE Domande SET testo = @testo, argomento = @argomento, RispostaA = @RispostaA, 
                RispostaB = @RispostaB, RispostaC = @RispostaC, RispostaD = @RispostaD, 
                RispostaCorretta = @RispostaCorretta, difficolta = @difficolta, 
                tempoRisposta = @tempoRisposta, meme = @meme, fonte = @fonte WHERE NumeroDomanda = @id", connection);

            command.Parameters.AddWithValue("@testo", domanda.Testo);
            command.Parameters.AddWithValue("@argomento", domanda.Argomento);
            command.Parameters.AddWithValue("@RispostaA", domanda.RispostaA);
            command.Parameters.AddWithValue("@RispostaB", domanda.RispostaB);
            command.Parameters.AddWithValue("@RispostaC", domanda.RispostaC);
            command.Parameters.AddWithValue("@RispostaD", domanda.RispostaD);
            command.Parameters.AddWithValue("@RispostaCorretta", domanda.RispostaCorretta);
            command.Parameters.AddWithValue("@difficolta", domanda.Difficolta);
            command.Parameters.AddWithValue("@tempoRisposta", domanda.TempoRisposta);
            command.Parameters.AddWithValue("@meme", domanda.Meme is null ? DBNull.Value : (object)domanda.Meme);
            command.Parameters.AddWithValue("@id", domanda.NumeroDomanda);
            command.Parameters.AddWithValue("@fonte", domanda.Fonte is null ? DBNull.Value : (object)domanda.Fonte);
            
            return command.ExecuteNonQuery() == 1;
        }

        public bool DoSaveOrUpdate(Domanda domanda)
        {
            Debug.WriteLine("DoSaveOrUpdate");
            if (DoRetrieveById(domanda.NumeroDomanda) == null)
            {
                return DoSave(domanda);
            }
            else
            {
                return DoUpdate(domanda);
            }
        }

        public bool DoDelete(Domanda domanda)
        {
            Debug.WriteLine("DoDelete");
            SqliteCommand command = new SqliteCommand("DELETE FROM Domande WHERE NumeroDomanda = @id", connection);
            command.Parameters.AddWithValue("@id", domanda.NumeroDomanda);
            return command.ExecuteNonQuery() == 1;
        }
    }
}
